### DISCUSSION TITLE
Multi Model messaging does not appear to be working correctly...

### DISCUSSION THREAD

**Discussion by nghtstr** [charmbracelet/bubbletea#560] [Q&A]
So, I have a main model, and two sub models. The `main` model is basically doing nothing more than displaying the two other models (`status` and `currentTest`) in the format that I want. So far, that is working great. Now, I have in the `main` model, I have the current state of the application, which includes information that the rest of the models will require. I got around this problem by passing that state as a part of a `tea.Cmd` message, which the two other models get. Now, the `currentTest` model has its own state listeners, and this is where things are breaking down. For whatever reason, even though I *know* that the secondary sate is being sent as a message (gotta love logging), my `currentTest` model isn't picking it up. Here is a code snippet:

```go
func (ct currentTestModel) sendCBState() tea.Msg {
	log.Println("Here 3")
	log.Printf("cbState: %T\n", ct.cbState)
	return ct.cbState
}

func (ct currentTestModel) sendGatheringState() tea.Msg {
	return ct.cbState.currentGatheringState
}

func (ct currentTestModel) Init() tea.Cmd {
	return ct.createDatabaseConnection
}

func (ct currentTestModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, DefaultKeyMap.Quit):
			return ct, tea.Quit
		}
	case tea.WindowSizeMsg:
		return ct, nil
	case state:
		sendState := false
		ct.currentState, sendState = ct.currentState.update(msg)
		ct.cbState.currentIndex = -1

		if sendState {
			log.Println("Here 2")
			return ct, tea.Batch(ct.sendState, ct.sendCBState)
		}

		return ct, ct.sendCBState
	case costbasisState:
		log.Println("Here")
		sendState := false
		ct.cbState, sendState = ct.cbState.update(msg)

		if sendState {
			return ct, tea.Batch(ct.sendCBState, ct.sendGatheringState)
		}
    ...
}
```

Also, here is the contents of my log file:
```
debug 2022/10/18 20:25:56 Here 2
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
debug 2022/10/18 20:25:56 Here 3
debug 2022/10/18 20:25:56 cbState: main.costbasisState
debug 2022/10/18 20:25:56 Here 2
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
debug 2022/10/18 20:25:56 Here 3
debug 2022/10/18 20:25:56 cbState: main.costbasisState
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
debug 2022/10/18 20:25:56 Here 3
debug 2022/10/18 20:25:56 cbState: main.costbasisState
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
debug 2022/10/18 20:25:56 Here 3
debug 2022/10/18 20:25:56 cbState: main.costbasisState
debug 2022/10/18 20:25:56 Should not have gotten here (Type: main.mainModel)
```

**Comment by erwin** [comment]
I was also struggling to figure out how to implement an application that held multiple models, until I found this example in the bubblezone codebase:

https://github.com/lrstanley/bubblezone/blob/master/examples/full-lipgloss/main.go

Perhaps I've missed something. It seems more simple to me than what you describe manually passing around `state`...

Basically inside of the `Update()` method for your main model, do something like:

```
	return m.propagate(msg), nil
```

Where `propagate` forwards event messages to the nested models.

Note that the `bubblezone` example is very simple though, it's not actually tracking the `cmd` returned from your sub-models, so if your sub-models generate events, you have to make sure to save them and append them with something like...


```
func (m MainModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
  util.LogF("%s", msg)
  var (
    cmd   tea.Cmd
    cmds  []tea.Cmd
  )
  // do MainModel's event handling...
  return m.propagate(msg, cmds)
}

func (m *MainModel) propagate(msg tea.Msg, cmds []tea.Cmd) (tea.Model, tea.Cmd) {
  var (
    cmd   tea.Cmd
  )
  switch msg := msg.(type) {
  default:
    m.status, cmd    = m.status.Update(msg)
    cmds            = append(cmds, cmd)
    m.currentTest, cmd = m.currentTest.Update(msg)
    cmds            = append(cmds, cmd)
  }
  return m, tea.Batch(cmds...)
}


```
